Add  categorization of language extensions

telser opened the pull request:
**Edit:** This [proposal](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0669-extension-categorization.md) has been accepted; the following discussion is mostly of historic interest

To be clear, if there is any controversy on a categorization the intent is to move it to the uncategorized section and leave it to another time. The goal is to get us through the biggest chunk possible of the categorization here without having to discuss too heavily on any one extension.

telser committed:
Add an initial categorization of language extensions

jvanbruegge commented:
Why have a second column for the stability of `-XNo...`? Isn't that always stable (as is the case here)?

s-and-witch commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -0,0 +1,224 @@
+| TemplateHaskell         |
+| TemplateHaskellQuotes   |
+| Trustworthy             |
+| TypeAbstractions        |
```
Why `TypeAbstractions` are uncategorized when `RequiredTypeArguments` are experimental?

telser replied:
In short, because I missed it. I'll add `TypeAbstractions`, the user's guide already says `Experimental` anyway.

s-and-witch commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -0,0 +1,224 @@
+| Extension              | Categorization | Negative Categorization |
+|------------------------|----------------|-------------------------|
+| ApplicativeDo          | Experimental   | Stable                  |
+| Arrows                 | Experimental   | Stable                  |
```
I don't like the presence of `Arrows` and `ApplicativeDo` here, or the word "experimental". These extensions aren't experimental, there is no experiment, they just have a bunch of long-standing bugs that someone may decide to solve at some point in the future

s-and-witch replied:
I'm objecting simply because of the implications. I like the "preview" term because it doesn't imply an experiment, old "preview" extension just means that it's stale/frozen in its preview form.

telser replied:
My stance is that we could spin on the exact terms and never make any progress on other things. As @adamgundry points out there is some compromise on any such terminology. As stated in the PR message, in order to make some progress on the majority, I think it's best to move controversial choices to uncategorized and revisit later.

telser replied:
I've moved both of these to uncategorized for the purposes of the this proposal.

telser committed:
Add paragraph about negative categorization

telser commented:
> Why have a second column for the stability of `-XNo...`? Isn't that always stable (as is the case here)?

In theory it isn't, so I've included it to be complete. An example could be `NoFieldSelectors`, might be considered `Experimental`. The proposal has been updated to have a couple of sentences to be more clear on this.

int-index commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -0,0 +1,234 @@
+| PolymorphicComponents             | Deprecated     | Stable                  |
+| Rank2Types                        | Deprecated     | Stable                  |
+| RecordPuns                        | Deprecated     | Stable                  |
+| TypeInType                        | Deprecated     | Stable                  |
```
Why is the "negative categorization" marked as "stable" for TypeInType? You get the warning either way

```
ghci> :set -XTypeInType
<no location info>: warning: [GHC-53692] [-Wdeprecated-flags]
    -XTypeInType is deprecated: use -XDataKinds and -XPolyKinds instead

ghci> :set -XNoTypeInType
<no location info>: warning: [GHC-53692] [-Wdeprecated-flags]
    -XNoTypeInType is deprecated: use -XNoDataKinds and -XNoPolyKinds instead
```

i.e. the entire extension flag is deprecated.

The same applies to

adamgundry commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -0,0 +1,234 @@
+| UndecidableSuperClasses |
+| Unsafe                  |
+
+## 3. Effect and Interactions
```
Please can we say explicitly what the effect is on the existing `-Wdeprecated-flags` warning (see https://github.com/ghc-proposals/ghc-proposals/pull/601#issuecomment-2103987867)?

adamgundry reviewed (commented):
Many thanks for pushing this forward @telser!

telser committed:
Address typos, formatting and other feedback

- Fixes typos and section header mismatches.

- Marks deprecated extensions as being deprecated in the negative case
too per feedback.

- Moves some extensions to uncategorized where there is some question on
them.

- Adds a mention of including `Deprecated` extensions in
`-Wdeprecated-flags` per feedback.

telser commented:
@adamgundry I believe I've addressed all the feedback you had. Thank you for the thorough look!

adamgundry commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -0,0 +1,240 @@
+
+#### 2.1.2 Deprecated Extensions
+
+The following are extensions that are to be categorized as `Deprecated` for both the enabled and `No` versions.
```
I still find it slightly confusing that it looks like the default (disabled) state is being deprecated in these cases, although I can see the argument for doing so. Perhaps we could add a sentence or two to emphasize that the deprecation applies to the *flag* `-XNo...` because there is no point in explicitly disabling these extensions, even though all the language editions include equivalent behaviour?

telser replied:
This is a great point. I'll add something for this.

adamgundry replied:
Thanks for clarifying this @telser, it looks good to me!

telser committed:
Pull request feedback

Add language about deprecation of No* flag but not behavior

Mark TypeAbstractions as experimental

bgamari commented:
There are a few extensions in the `Stable` list here which I would say shouldn't belong:

* `TypeFamilies` (and similarly `TypeFamilyDependencies`) is certainly not well-enough specified enough to be considered to be stable
* I am a bit concerned that `UnliftedFFITypes`'s semantics are too unsafe to 
* `UnliftedDatatypes` seems like it may be a bit young to be considered to be stable; we might consider holding off a bit
* While `TransformListComp` is not likely to change at this point, it has very little usage in the wild (only 7 usages on Hackage); it is not clear to me that there is much benefit in declaring it to be stable at this point.
* `MagicHash` has known parsing pecularities that we might consider cleaning up in the future. Perhaps it would be best to leave it out for now.

simonpj commented:
> While TransformListComp is not likely to change at this point, it has very little usage in the wild (only 7 usages on Hackage); it is not clear to me that there is much benefit in declaring it to be stable at this point.

I agree -- but perhaps we can add a "Reasoning" column in the table, to explain the rationale for the choice.  (We don't have to fill it in for every extension.)  That would save iterating the same argument in 3 yrs time.

The rationale for `TypeFamilies` is very different. It is *very* heavily used.

maralorn commented:
The problem is how we apply our stability criteria here: We would probably conform with all our formal rules by defining `TypeFamilies` as experimental in one of the next releases and then rewrite it/throw it out.
But in truth that would be a massive breaking change which is most likely a non-starter and would not conform to the spirit of our stability rules. So I fear even though an extension should really be considered experimental because of any number of design concerns, if it is years old and in heavy use it is defacto stable. Maybe we can loosen that law for extensions which have "experimental" written on them loud and clearly from the beginning like we do now, but even that could fail …
My feeling for stuff like TypeFamilies is the best we can do is offer a better replacement under a new extension and then mark it legacy (or maybe even deprecated far down the line, but it seems unlikely.)
I fear this whole classification hinges on how we stand on that. I don’t think it makes sense to mark anything experimental, if we are not prepared to break it.

I guess the intention of this proposal is to not make a decision about this and postpone that discussion?

If so there are other extensions with not completely settled designs which should probably not classified as stable just yet. I nominate OverloadedStrings and ViewPatterns for that …

mpickering commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -0,0 +1,244 @@
+| NumericUnderscores         | Stable         | Stable                  |
+| OverloadedLabels           | Stable         | Stable                  |
+| OverloadedLists            | Stable         | Stable                  |
+| OverloadedRecordDot        | Stable         | Stable                  |
```
My impression is that `OverloadedRecordDot` (or perhaps just record extensions in general) have received quite a lot of changes and proposals recently. Perhaps @adamgundry knows more?

adamgundry replied:
I think `OverloadedRecordDot` has been fairly stable. Recently I generalised it (well, `HasField`) to support representation polymorphism, and there's a proposal to slightly widen the syntax at #668, but I don't expect any of that to break existing code.

The situation is slightly different with some of the other record extensions, e.g. `OverloadedRecordUpdate` is very unstable, while `DuplicateRecordFields` and `NoFieldSelectors` are somewhat in the middle, but that's reflected in the current categorisation.

telser replied:
Either way, I've moved it to uncategorized for now.

adamgundry commented:
I think the bar to be met for `Stable` is not "are we completely happy with the design of this extension" but "is this extension well-specified enough that we can avoid unexpected breaking changes". So for extensions like `TupleSections` or `OverloadedStrings` my (admittedly ill-informed) assumption would be that even if we later want to do something different, it wouldn't be that hard to control the new thing under a new extension and avoid breaking existing code.

`TypeFamilies` is a bit tricky, because while it's very widely used and is unlikely to change in a way that leads to significant breakage in simple cases (we'd instead introduce a new extension), there are dark corners of the constraint solver that may be hard to keep completely stable. It's not obvious whether we should try to assign responsibility there to specific extensions, because often it is in the interactions between fancy type-level features that the hard cases arise.

phadej commented:
@adamgundry I think it's worth separating stable from "unlikely to be extended" "likely to be extended", the latter doesn't fit well to being added to GHC20xx language editions (if we want treat them as "standards", i.e. all GHC are able to compile these programs).

telser committed:
Update categorizations based on feedback.

Seven more extensions had some dispute on stability so leaving them
uncategorized.

They are:
- OverloadedRecordDot
- TransformListComp
- TupleSections
- TypeFamilies
- TypeFamilyDependencies
- UnliftedDatatypes
- UnliftedFFITypes

telser commented:
> There are a few extensions in the `Stable` list here which I would say shouldn't belong:
> 
>     * `TypeFamilies` (and similarly `TypeFamilyDependencies`) is certainly not well-enough specified enough to be considered to be stable
> 
>     * I am a bit concerned that `UnliftedFFITypes`'s semantics are too unsafe to
> 
>     * `UnliftedDatatypes` seems like it may be a bit young to be considered to be stable; we might consider holding off a bit
> 
>     * While `TransformListComp` is not likely to change at this point, it has very little usage in the wild (only 7 usages on Hackage); it is not clear to me that there is much benefit in declaring it to be stable at this point.

I've adjusted these to be uncategorized for now. Thanks!

telser commented:
> Personally I feel like there is some conflict between TupleSections and the potential for future trailing comma proposals.
> 
> However I have to admit that since TupleSections has been added to GHC202[14] already, perhaps this is not enough of a reason to omit it from the list of stable extensions.

Moved it out of the stable list. Though I would like to think, especially for GHC2024 extensions that conflicts would be considered very carefully.

telser committed:
Mark two more extensions uncategorized

- OverloadedStrings
- ViewPatterns

telser commented:
> I don’t think it makes sense to mark anything experimental, if we are not prepared to break it.

This is a low pass filter for stable. I'd say start the other way around "experimental" is the default state and it doesn't make sense to mark otherwise unless prepared to *not* break it.

> 
> I guess the intention of this proposal is to not make a decision about this and postpone that discussion?

That is very much the idea here. There are lots of extensions that raise any number of questions. So let's get the, hopefully large, set of easier to agree on done. That shows some progress and we can figure out the harder cases in additional passes.

> 
> If so there are other extensions with not completely settled designs which should probably not classified as stable just yet. I nominate OverloadedStrings and ViewPatterns for that …

Moved both of those to uncategorized.

phadej commented on proposals/0000-initial-extension-categorization.md:
```diff
@@ -207,10 +200,16 @@ The following are the extensions that are uncategorized as of this proposal. The
 | TemplateHaskellQuotes   |
+| TransformListComp       |
 | Trustworthy             |
+| TupleSections           |
```
It's somewhat surprising to see `TupleSections` in an uncategorized category, as it is included in `GHC2021` and `GHC2024`. Similarly `MonoLocalBinds`.

I'd hope they will be next on the list to discuss.

phadej replied:
FWIW,

```
Prelude> :set -XNoRelaxedPolyRec
<no location info>: warning: [GHC-53692] [-Wdeprecated-flags]
    -XNoRelaxedPolyRec is deprecated: You can't turn off RelaxedPolyRec any more
```

I guess it arguable that it may not be stable as whole extension flag could be removed. But it's an interesting special case.

telser commented:
@adamgundry I think this is ready to go to the committee now.

maralorn commented:
Shepherd here. This is not a trivial proposal so I will need a few days to evaluate it in more detail before making a recommendation.

maralorn reviewed (dismissed):
Let me start with a big thank you, Trevis. I think this chain of proposals is an important step for transparency to our users and as a guideline to this committee.

I will certainly recommend acceptance to the committee. Something like this is the logical and sensible follow up to #601. (I however hope that we can find an effective way to classify all our extensions on a sensible time scale.)

That being said, besides some minor points and before we scrutinize the concrete list of extensions further I have two major points to discuss. Both of these points are not about classification, but concern the effect of the classifications and have not been addressed by #601.

1. Section 2 starts with

    > The entirety of the changes proposed are categorizations of extensions.

    but then Section 3 says

    > Extensions that are marked as `Deprecated` must be added to `-Wdeprecated-flags` to emit an appropriate warning on their use.

    In my opinion this should be part of "2 Proposed Change Specifiation".
    Also the the start of that section should clearly state that this proposal mainly categorizes extensions, but also changes (or if you want "clarifies") the categorization framework. #601 decided after much debate to not affect flags at all, so this point might be contentious.
    The proposal should also specify, whether all deprecated extensions should emit this warning or if it only mandates this for the ones marked deprecated here. (I imagine the later does not make a lot of sense, but it drives home the point, that this is a policy change.)

2. The whole negation stuff is muddying the waters. That is a consequence from the user guide being weird and #601 being a bit vague on this. My reading of #601 is that the categories are for language extensions and my reading of the user guide is that `FieldSelectors` is a language extension but `NoFieldSelectors` isn’t. That’s a bit unfortunate. Imagine the introduction of `NoPunning`, which would be Experimental, but `Punning` would be stable. Putting `Status: Stable` under `Punning` in the user guide would be unhelpful and putting `Status: Experimental` under `Punning` would be wrong. I think we should fix this. Not just out of principle but because we basically need to figure this out to decide how it looks in the user guide. Maybe someone has a smart idea for this? I can’t quite wrap my head around. Please chime in if you have an opinion about this.

    My intuition is quite clear, we want to say `CPP is Stable, NoFieldSelectors is Stable, Rank2Types is Deprecated, DeepSubsumption is Legacy, LinearTypes is Experimental, NoPunning is Experimental`. (just examples) Also I can’t come up with any situation where the other side is not `Stable` in some sense. `Deprecated` seems to be an exception but I think that is because that classification classifies flags and not extensions.

    The best idea I can currently come up with is changing the categorization framework to "Language extensions (or their negation but not both) will be classified into the following categories:". I think this would be good because the framework already highlights quite a number of category transitions and duplicating that for both sides of an extension seems quite confusing and I hope it is unnecessarily complex. We might want to clarify that the negation of the classified setting is always implicitly stable and in regard to my first point that for Deprecated extensions any explicit toggling triggers a warning.

Lastly I suggest finding an equal title for PR and document and suggest "Initial Categorization of Language Extensions"

adamgundry commented:
@maralorn The negative extensions issue (and the interaction with deprecation of flags) is indeed a tricky one. As you describe, it does sound like a plausible policy simplification to say that either
 * one polarity of an extension is `Experimental`/`Stable`/`Legacy` and the other is `Stable`, or
 * both polarities are `Deprecated`.

Regarding `-Wdeprecated-flags`, I would hope this is less controversial than the discussion in #601 around introducing new warnings. The point is that we already have `-Wdeprecated-flags`, and it doesn't really make sense to have two classes of deprecated extensions (those considered `Deprecated` and those warned about by `-Wdeprecated-flags`), instead we should collapse the distinction. But I agree the proposal should be up front about this.

Perhaps as part of this PR we should revise #601 with these policy clarifications, while the new document can contain the concrete classification? (Let's make sure we have consensus on the policy first though.)

maralorn commented:
@telser Please contact me if I can help you in anyway. I’d e.g. be happy to jump on a call, if you need someone to brainstorm.

Until then I will change the state of this proposal, but I will jump back into action as soon as there is any movement.

telser committed:
Make small amendments given feedback

- Explicitly merge the notion of Deprecation with the warning flag

- Be explicit about extension negation handling

telser committed:
More small amendments

Adjusting for conversations at Zurihac

telser committed:
Mention deprecation warning flag earlier

telser committed:
Rename proposal file to match title change

maralorn reviewed (commented):
I recommended acceptance to the committee and we are now in discussion.

Because of vacation time and the large number of extensions to review I expect this to take at least 4 weeks.

telser commented:
> I recommended acceptance to the committee and we are now in discussion.
> 
> Because of vacation time and the large number of extensions to review I expect this to take at least 4 weeks.

Thank you for the update and timeline @maralorn !

simonpj commented:
**Attention all GHC users**.  This proposal categorises GHC extensions into

- stable
- deprecated
- legacy
- experimental
- uncategorised

This classification should be of interest to everyone, so you might want to take a final look because the proposal has now gone to the committee for decision.

The intent is the anything controversial goes into "uncategorised" so that we can make progress with uncontroversial classifications.

phadej commented:
I'm also surprised to see `OverloadedLists` in stable, but not `OverloadedStrings`. I understand that there's discussions of unsatisfactory state of `OverloadedStrings`, but I would raise concern that `OverloadedLists` is at least as much unsatisfactory. If the question is about partiality, then `OverloadedLists` cannot be used to construct `NonEmpty`, but it would be nice to write `[1, 2, 3] :: NonEmpty Int` and make GHC fail at compile time if I try to do `[] :: NonEmpty Int`.

telser committed:
Move GADTs and OverloadedLists to uncategorized

simonpj commented:
Good points Simon.

I can't even remember exactly what `RelaxedPolyRec` does!  It is now definitely deprecated, and you can't turn it off.  

I think it means that in a mutually recursive group, { f = ...g... ; g= .. f .. }, if you give a type signature for just one (say f) we'll typecheck `g` usign that type signature and generalise it.  So one type sig may be enough to break the knot. 

GHC has done this *forever*.  It's so stable you can't turn it off, and is now deprecated.

telser committed:
Add ExplicitNamespaces to uncategorized

phadej commented:
@adamgundry 

> I think ultimately GADTs plus MonoLocalBinds should both be Stable (but there's an argument for making NoMonoLocalBinds into Legacy). I

Except by default it is `NoMonoLocalBinds`, and having Legacy extension enabled by default is probably why it resists being categorised.

int-index commented:
`NoListTuplePuns` should be marked experimental, while `ListTuplePuns` is stable.

telser commented:
> I don't really see the point of this dual categorisation presentation. Unless I'm mistaken the “Off” classification is a function of the “On” classification. So you don't need this second column, you can just document the function instead. What am I missing?

There has already been *much* confusion when the 'No' variant is the feature. One example being `NoFieldSelectors`, though there are others. I would much rather be explicit and clear here.

aspiwack commented:
I believe @int-index comment was meant as an answer to my question, and as support to the two-column categorisation, rather than as a request for your proposal.

As for me, I don't much care either way. It had caught my eye, so I said my piece. And now I'll leave the rest of you to decide what to do.

jappeace commented:
Hi @VitWW please note that we've discussed these buckets for years at this point. If you wish to add more of these we can do this in a future proposal, or please join the Haskell Stability Working Group if you're interested in this.

jappeace commented:
@VitWW  I encourage you to open another proposal to address this.

the categories were already accepted here: https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0601-extension-lifecycle-framework.md

this proposal isn't about that definition.

telser commented:
@maralorn Any updates on this?

maralorn committed:
Accept extension-categorization proposal

maralorn commented:
I apologize for the delay. I set a long deadline and tricked myself into thinking it was even longer.

The proposal is now accepted. Thank you Trevis for pushing this through!
